source/base/mu_parser_internal.cc

```diff
@@ -0,0 +1,220 @@
+
+#include <deal.II/base/mu_parser_internal.h>
+
+
+DEAL_II_NAMESPACE_OPEN
+
+
+
+#ifdef DEAL_II_WITH_MUPARSER
+
+namespace internal
+{
+  int
+  mu_round(double val)
+  {
+    return static_cast<int>(val + ((val >= 0.0) ? 0.5 : -0.5));
+  }
+
+  double
+  mu_if(double condition, double thenvalue, double elsevalue)
+  {
+    if (mu_round(condition))
+      return thenvalue;
+    else
+      return elsevalue;
+  }
+
+  double
+  mu_or(double left, double right)
+  {
+    return (mu_round(left)) || (mu_round(right));
+  }
+
+  double
+  mu_and(double left, double right)
+  {
+    return (mu_round(left)) && (mu_round(right));
+  }
+
+  double
+  mu_int(double value)
+  {
+    return static_cast<double>(mu_round(value));
+  }
+
+  double
+  mu_ceil(double value)
+  {
+    return std::ceil(value);
+  }
+
+  double
+  mu_floor(double value)
+  {
+    return std::floor(value);
+  }
+
+  double
+  mu_cot(double value)
+  {
+    return 1.0 / std::tan(value);
+  }
+
+  double
+  mu_csc(double value)
+  {
+    return 1.0 / std::sin(value);
+  }
+
+  double
+  mu_sec(double value)
+  {
+    return 1.0 / std::cos(value);
+  }
+
+  double
+  mu_log(double value)
+  {
+    return std::log(value);
+  }
+
+  double
+  mu_pow(double a, double b)
+  {
+    return std::pow(a, b);
+  }
+
+  double
+  mu_erfc(double value)
+  {
+    return std::erfc(value);
+  }
+
+  // returns a random value in the range [0,1] initializing the generator
+  // with the given seed
+  double
+  mu_rand_seed(double seed)
+  {
+    static Threads::Mutex       rand_mutex;
+    std::lock_guard<std::mutex> lock(rand_mutex);
+
+    static boost::random::uniform_real_distribution<> uniform_distribution(0,
+                                                                           1);
+
+    // for each seed an unique random number generator is created,
+    // which is initialized with the seed itself
+    // we could use std::mt19937 but doing so results in compiler-dependent
+    // output.
+    static std::map<double, boost::random::mt19937> rng_map;
+
+    if (rng_map.find(seed) == rng_map.end())
+      rng_map[seed] = boost::random::mt19937(static_cast<unsigned int>(seed));
+
+    return uniform_distribution(rng_map[seed]);
+  }
+
+  // returns a random value in the range [0,1]
+  double
+  mu_rand()
+  {
+    static Threads::Mutex                             rand_mutex;
+    std::lock_guard<std::mutex>                       lock(rand_mutex);
+    static boost::random::uniform_real_distribution<> uniform_distribution(0,
+                                                                           1);
+    static boost::random::mt19937                     rng(
+                          static_cast<unsigned long>(std::time(nullptr)));
+    return uniform_distribution(rng);
+  }
+
+  // helper function to determine the past-the-end pointer of an array
+  template<typename T, std::size_t N>
+  T* past_the_end_ptr(T (&ra)[N])
```

Comment:
There is `std::end()` defined in header `<iterator>` that provides the same functionality.

Reply:
Class removed.